scipy/sparse/tests/test_base.py

```diff
@@ -3972,6 +3972,28 @@ class TestLIL(sparse_test_class(minmax=False)):
     spmatrix = lil_matrix
     math_dtypes = [np.int_, np.float_, np.complex_]
 
+    def test_constructor1(self):
+        # (data, rows), shape=(M, N) format
+        data1 = [[1], [2], [3]]
+        rows1 = [[0], [1], [2]]
```

Comment:
For slightly more coverage, try using some non-list types for these. For example:

```
data1 = (np.ones(1), (2,), {3})
```

Reply:
Seems `isintlike` does not like mixed type tuple. 

Here is the test log (numpy v1.19.2)

```python
======================================================================== FAILURES ========================================================================
_______________________________________________________________ TestLIL.test_constructor1 ________________________________________________________________
/home/oreki_clr/miniconda3/envs/scipydev/lib/python3.8/site-packages/numpy/core/fromnumeric.py:3108: in ndim
    return a.ndim
E   AttributeError: 'tuple' object has no attribute 'ndim'
        a          = (array([1.]), (2,), {3})

During handling of the above exception, another exception occurred:
scipy/sparse/tests/test_base.py:3981: in test_constructor1
    B = lil_matrix((data1, rows1), shape=(3, 3))
        A          = array([[1, 0, 0],
       [0, 2, 0],
       [0, 0, 3]])
        data1      = (array([1.]), (2,), {3})
        rows1      = [[0], [1], [2]]
        self       = <scipy.sparse.tests.test_base.TestLIL object at 0x7fc6202bcdc0>
scipy/sparse/lil.py:109: in __init__
    if isshape(arg1):
        arg1       = ((array([1.]), (2,), {3}), [[0], [1], [2]])
        copy       = False
        dtype      = None
        self       = <[AttributeError('nnz not found') raised in repr()] lil_matrix object at 0x7fc6202bcf40>
        shape      = (3, 3)
scipy/sparse/sputils.py:219: in isshape
    if isintlike(M) and isintlike(N):
        M          = (array([1.]), (2,), {3})
        N          = [[0], [1], [2]]
        nonneg     = False
        x          = ((array([1.]), (2,), {3}), [[0], [1], [2]])
scipy/sparse/sputils.py:192: in isintlike
    if np.ndim(x) != 0:
        x          = (array([1.]), (2,), {3})
<__array_function__ internals>:5: in ndim
    ???
        args       = ((array([1.]), (2,), {3}),)
        kwargs     = {}
        relevant_args = ((array([1.]), (2,), {3}),)
/home/oreki_clr/miniconda3/envs/scipydev/lib/python3.8/site-packages/numpy/core/fromnumeric.py:3110: in ndim
    return asarray(a).ndim
        a          = (array([1.]), (2,), {3})
/home/oreki_clr/miniconda3/envs/scipydev/lib/python3.8/site-packages/numpy/core/_asarray.py:83: in asarray
    return array(a, dtype, copy=False, order=order)
E   numpy.VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
        a          = (array([1.]), (2,), {3})
        dtype      = None
        order      = None
================================================================ short test summary info =================================================================
FAILED scipy/sparse/tests/test_base.py::TestLIL::test_constructor1 - numpy.VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences ...
=================================================================== 1 failed in 0.39s ====================================================================

```

Reply:
This is actually a good problem to catch with testing, then! The problem is distinguishing between calling `lil_matrix(shape)` and `lil_matrix((data, rows))`, as they're both 2-item tuples.  The current `isintlike()` check calls `np.ndim()`, which in turn tries to coerce its input to an ndarray, which we probably want to avoid if possible.

Perhaps we should avoid the `isshape()` call and  check if the entries in the tuple are sequences with `issequence()` instead. This wouldn't allow the set `{3}`, but I'm okay with not supporting sets as inputs.

Reply:
Yeah, It would be best to avoid `isshape()` call.
- Are you suggesting to check entires of `data` and `rows` for sequence with `issequence()` ? If so then we will need to first check if entires (`data` & `rows`) in tuple `arg1` are iterable or not also.
  ```python
    elif len(arg1) == 2 and all(
            isinstance(a, (list, tuple, np.ndarray)) for a in arg1
            ) and all(all(issequence(s) for s in a) for a in arg1):
        # (data, rows) format 
  ```
- Also we can just replace call to `isshape(arg1)` with `len(arg1) == 2 and issequence(arg1)`  as subsequent call to `check_shape()` will validate whether tuple is valid shape or not (In this case even sets would be allowed as inputs)
  ```python
     # if isshape(arg1):
     if len(arg1) == 2 and issequence(arg1):
         # shape
         M, N = arg1
         self._shape = check_shape((M, N))
         ...
     elif len(arg1) == 2:
         # (data, rows) format
         (data, rows) = arg1
         ...
     else:
         raise TypeError('unrecognized lil_matrix constructor usage')
  ```